### ISSUE TITLE
[QUESTION] Handling Unlabeled and Background Classes in Mask2Former for Custom Dataset

### ISSUE THREAD

**Issue by xi-loong** [lightly-ai/lightly-train#403]
### 🤔 What’s your question?

I want to train a model on a dataset from https://arxiv.org/html/2507.16855v1, where the labels are defined as: class 0 for unlabeled pixels (due to uncertain categories), class 1 for background, and classes 2-16 for foreground. However, when I examined the Mask2Former loss code, it seems to treat the ignore_index and background as the same category. Does this mean all ignore_index pixels would be predicted as background? This appears incompatible with my dataset, where unlabeled (class 0) and background (class 1) are distinct. How should I handle these classes to ensure proper training with Mask2Former?

**Comment by stegmuel** [comment]
Hi,

Great to see an application to histopathology!
If you set the data dict as follows, the training should behaves as expected:
```
        data={
            "train": {
                "images": "my_data_dir/train/images",   # Path to training images
                "masks": "my_data_dir/train/masks",     # Path to training masks
            },
            "val": {
                "images": "my_data_dir/val/images",     # Path to validation images
                "masks": "my_data_dir/val/masks",       # Path to validation masks
            },
            "classes": {                                # Classes in the dataset                    
                0: "unlabeled",
                1: "background",
                2: "tumor cells",
                3: "reactive epithelium",
                # ...
            },
            # Optional, classes that are in the dataset but should be ignored during
            # training.
            "ignore_classes": [0], 
        },
```
In that setting, no loss is enforced on pixels corresponding to the class 0 during training and it will also never be predicted at inference time. the `background` class is treated as any other class. Is that the desired behavior?

**Comment by stegmuel** [comment]
FYI, in the past we were able to load [Phikon-v2](https://arxiv.org/pdf/2409.09173) weights in our framework for segmentation finetuning. If you need help loading some of the more recent SOTA weights ([H-Optimus-1](https://huggingface.co/bioptimus/H-optimus-1), _etc_.), let me know.

**Comment by xi-loong** [comment]
> Hi,
> 
> Great to see an application to histopathology! If you set the data dict as follows, the training should behaves as expected:
> 
> ```
>         data={
>             "train": {
>                 "images": "my_data_dir/train/images",   # Path to training images
>                 "masks": "my_data_dir/train/masks",     # Path to training masks
>             },
>             "val": {
>                 "images": "my_data_dir/val/images",     # Path to validation images
>                 "masks": "my_data_dir/val/masks",       # Path to validation masks
>             },
>             "classes": {                                # Classes in the dataset                    
>                 0: "unlabeled",
>                 1: "background",
>                 2: "tumor cells",
>                 3: "reactive epithelium",
>                 # ...
>             },
>             # Optional, classes that are in the dataset but should be ignored during
>             # training.
>             "ignore_classes": [0], 
>         },
> ```
> 
> In that setting, no loss is enforced on pixels corresponding to the class 0 during training and it will also never be predicted at inference time. the `background` class is treated as any other class. Is that the desired behavior?


Hello, and thank you for your response. This is indeed the behavior I was expecting.

However, when training with slightly_train's DINOv3EoMTSemanticSegmentationTrain, I encountered an issue where the loss value was abnormally high (around 80).

After reviewing the source code of slightly_train, I noticed that the MaskClassificationLoss doesn't seem to properly implement the loss calculation considering the ignore_classes. Specifically, DINOv3EoMTSemanticSegmentationTrain sets the ignore_classes as the ignored_index, then performs Hungarian matching, and finally computes BCEWithLogitsLoss and dice_loss.

The problem is that BCEWithLogitsLoss does not natively support an ignore_index parameter like CrossEntropyLoss does. I separately tested BCEWithLogitsLoss, and when I set certain pixels in the ground truth labels to ignore_index (e.g., -100), the resulting loss was very large (around 60), which is on the same order of magnitude as the high loss I observed during full training.

Therefore, I personally suspect that the current MaskClassificationLoss, when used for semantic segmentation models, might not actually support the intended ignoring functionality for ignore_classes.

Of course, I'm not entirely sure if this assessment is accurate. I plan to proceed with the training first and then verify the model's actual performance by evaluating the final mIOU metric.

**Comment by stegmuel** [comment]
Hi,

Sorry for the delay. 
In the end how did the mIoU looked like?
In the dataset class we compute [binary masks for each valid class](https://github.com/lightly-ai/lightly-train/blob/d55b51325a7e7f1bfee63bb816c10515f801b6ab/src/lightly_train/_data/mask_semantic_segmentation_dataset.py#L119) and no binary mask is created for the `ignore_index`class. 
The dataset class still yields standard segmentation masks with pixels potentially having the `ignore_index` value. These masks are only used for mIoU calculation not losses and the metric takes as input the `ignore_index` [here](https://github.com/lightly-ai/lightly-train/blob/d55b51325a7e7f1bfee63bb816c10515f801b6ab/src/lightly_train/_task_models/dinov3_eomt_semantic_segmentation/train_model.py#L225).
The BCEWithLogitsLoss does not take as parameter an `ignore_index` because it is used for binary classification, here background/foreground classification. In the mask classification paradigm, the segmentation and classification tasks are disentangled. The [`loss_masks`](https://github.com/lightly-ai/lightly-train/blob/d55b51325a7e7f1bfee63bb816c10515f801b6ab/src/lightly_train/_task_models/dinov3_eomt_semantic_segmentation/mask_loss.py#L79) ensures binary predicted masks are well aligned to ground truth masks (assigned through Hungarian matching). The [`loss_classes`](https://github.com/lightly-ai/lightly-train/blob/d55b51325a7e7f1bfee63bb816c10515f801b6ab/src/lightly_train/_task_models/dinov3_eomt_semantic_segmentation/mask_loss.py#L80) ensures queries corresponding to the different predicted masks are correctly classified.

**Comment by xi-loong** [comment]
Hello,

Thank you for your follow-up. I'm currently conducting comparative experiments and have observed a significant gap between training and validation mIoU. However, the validation mIoU has reached 51, which confirms that the approach is working. Thank you for the code example you provided earlier.

I've identified that the main reason for the large mIoU gap was the validation set's image preprocessing - the `SmallestMaxSize` operation was resizing original large images to 518px, causing much more aggressive scaling in validation compared to training. After fixing this preprocessing inconsistency, I'm proceeding with the following experiments:

Based on my analysis of Mask2Former's loss implementation, I believe there might be some limitations in how it handles unannotated areas in this dataset. Since the unannotated regions are quite large and may potentially contain tumor areas, the current approach that computes loss on these unannotated regions (essentially requiring all masks to output 0 in unannotated areas) might not be optimal.

Therefore, I plan to modify the `sample_points` function in Mask2Former to only sample points from non-unannotated regions (i.e., background and foreground areas) when calculating dice loss and sigmoid cross-entropy loss. This approach seems more reasonable as it won't force the model to output 0 in unannotated regions.

I will modify the lightly_train code to implement the point sampling strategy described above, and compare it with the unmodified version to validate the effectiveness of this modification.

Once these experiments are complete, I'll share the comparative results to demonstrate whether this approach provides better handling of unannotated regions during training.
